x-pack/filebeat/input/netflow/decoder/v9/lru.go

```diff
@@ -0,0 +1,164 @@
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+type eventWithMissingTemplate struct {
+	key       SessionKey
+	entryTime time.Time
+}
+
+type pendingEventsHeap []eventWithMissingTemplate
+
+func (h pendingEventsHeap) Len() int {
+	return len(h)
+}
+
+func (h pendingEventsHeap) Less(i, j int) bool {
+	return h[i].entryTime.Sub(h[j].entryTime) < 0
+}
+
+func (h pendingEventsHeap) Swap(i, j int) {
+	h[i], h[j] = h[j], h[i]
+}
+
+func (h *pendingEventsHeap) Push(x any) {
+	v, ok := x.(eventWithMissingTemplate)
+	if ok {
+		*h = append(*h, v)
+	}
+}
+
+func (h *pendingEventsHeap) Pop() any {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1 : n-1]
+	return x
+}
+
+type pendingTemplatesCache struct {
+	mtx     sync.RWMutex
+	wg      sync.WaitGroup
+	isEmpty atomic.Bool
+	hp      pendingEventsHeap
+	started bool
+	events  map[SessionKey][]*bytes.Buffer
+}
+
+func newPendingTemplatesCache() *pendingTemplatesCache {
+	cache := &pendingTemplatesCache{
+		events: make(map[SessionKey][]*bytes.Buffer),
+		hp:     pendingEventsHeap{},
+	}
+	return cache
+}
+
+// GetAndRemove returns all events for a given session key and removes them from the cache
+func (h *pendingTemplatesCache) GetAndRemove(key SessionKey) []*bytes.Buffer {
+	if h.isEmpty.Load() {
+		return nil
+	}
+
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+	events, ok := h.events[key]
+	if !ok {
+		return nil
+	}
+	delete(h.events, key)
+	h.isEmpty.Store(len(h.events) == 0)
+	return events
+}
+
+// Add adds an event to the pending templates cache
+func (h *pendingTemplatesCache) Add(key SessionKey, events *bytes.Buffer) {
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+
+	h.events[key] = append(h.events[key], events)
+	h.hp.Push(eventWithMissingTemplate{key: key, entryTime: time.Now()})
+	h.isEmpty.Store(false)
+}
+
+// start starts the pending templates cache cleaner
+func (h *pendingTemplatesCache) start(done <-chan struct{}, cleanInterval time.Duration, removalThreshold time.Duration) {
+	h.mtx.Lock()
+	if h.started {
+		h.mtx.Unlock()
+		return
+	}
+	h.started = true
+	h.mtx.Unlock()
+
+	h.wg.Add(1)
+	go func(n *pendingTemplatesCache) {
+		defer n.wg.Done()
+		timer := time.NewTimer(cleanInterval)
+		defer timer.Stop()
+
+		for {
+			select {
+			case <-timer.C:
+				h.cleanup(removalThreshold)
+				timer.Reset(cleanInterval)
+			case <-done:
+				return
+			}
+		}
+	}(h)
+}
+
+func (h *pendingTemplatesCache) cleanup(removalThreshold time.Duration) {
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+
+	if len(h.hp) == 0 {
+		// lru is empty do not proceed further
+		return
+	} else if len(h.events) == 0 {
+		// all pending events have been cleaned by GetAndRemove
+		// thus reset lru since it is not empty (look above) and continue
+		h.hp = pendingEventsHeap{}
+		return
+	}
+
+	hp := &h.hp
+	now := time.Now()
+	for {
+		v := heap.Pop(hp)
+		c, ok := v.(eventWithMissingTemplate)
+		if !ok {
+			// weirdly enough we should never get here
+			continue
+		}
+		if now.Sub(c.entryTime) < removalThreshold {
```

Comment:
nit: maybe ```time.Before```

Reply:
I did switch to that, thanks for the proposal. Feel free to unresolve if you spot anything else